src/software/ai/hl/stp/tactic/defense_shadow_enemy_tactic.cpp

```diff
@@ -0,0 +1,122 @@
+#include "software/ai/hl/stp/tactic/defense_shadow_enemy_tactic.h"
+
+#include "software/ai/hl/stp/action/move_action.h"
+#include "software/ai/hl/stp/action/stop_action.h"
+#include "software/ai/hl/stp/evaluation/calc_best_shot.h"
+#include "software/ai/hl/stp/evaluation/robot.h"
+#include "software/ai/hl/stp/tactic/tactic_visitor.h"
+#include "software/util/parameter/dynamic_parameters.h"
+
+DefenseShadowEnemyTactic::DefenseShadowEnemyTactic(
+    const Field &field, const Team &friendly_team, const Team &enemy_team,
+    const Ball &ball, bool ignore_goalie, double shadow_distance,
+    bool enemy_team_can_pass, bool loop_forever)
+    : Tactic(loop_forever),
+      field(field),
+      friendly_team(friendly_team),
+      enemy_team(enemy_team),
+      ball(ball),
+      ignore_goalie(ignore_goalie),
+      shadow_distance(shadow_distance),
+      enemy_team_can_pass(enemy_team_can_pass)
+{
+}
+
+std::string DefenseShadowEnemyTactic::getName() const
+{
+    return "Defense Shadow Enemy Tactic";
+}
+
+void DefenseShadowEnemyTactic::updateWorldParams(const Field &field,
+                                                 const Team &friendly_team,
+                                                 const Team &enemy_team, const Ball &ball)
+{
+    this->field         = field;
+    this->friendly_team = friendly_team;
+    this->enemy_team    = enemy_team;
+    this->ball          = ball;
+}
+
+void DefenseShadowEnemyTactic::updateControlParams(
+    const Evaluation::EnemyThreat &enemy_threat)
+{
+    this->enemy_threat = enemy_threat;
+}
+
+double DefenseShadowEnemyTactic::calculateRobotCost(const Robot &robot,
+                                                    const World &world)
+{
+    if (!enemy_threat)
+    {
+        return 0;
+    }
+    // Prefer robots closer to the enemy being shadowed
+    // We normalize with the total field length so that robots that are within the field
+    // have a cost less than 1
+    double cost = (robot.position() - enemy_threat->robot.position()).len() /
+                  world.field().totalXLength();
+    return std::clamp<double>(cost, 0, 1);
+}
+
+void DefenseShadowEnemyTactic::calculateNextIntent(IntentCoroutine::push_type &yield)
+{
+    MoveAction move_action = MoveAction();
+    StopAction stop_action =
+        StopAction(StopAction::ROBOT_STOPPED_SPEED_THRESHOLD_DEFAULT, true);
+    do
+    {
+        if (!enemy_threat)
+        {
+            yield(stop_action.updateStateAndGetNextIntent(*robot, false));
+        }
+
+        Robot enemy_robot                   = enemy_threat->robot;
+        std::vector<Robot> robots_to_ignore = {*robot};
+
+        if (ignore_goalie && friendly_team.goalie())
+        {
+            robots_to_ignore.emplace_back(*friendly_team.goalie());
+        }
+
```

Comment:
we're not using `enemy_can_pass` in here?

Reply:
good catch, wasn't used anywhere. Removed.

Reply:
Wasn't that an important part of strategy used at competition?

Reply:
Apparently not part of this tactic. This tactic is very similar to the regular `ShadowEnemy` but has ball-stealing capabilities instead of shadowing based on a pass. This is stuff we'll want to re-evaluate and likely combine once everything is ported.